pkg/operator/kube_cloud_config/controller.go

```diff
@@ -0,0 +1,143 @@
+// as source for configuration on clouds/platform.
+type KubeCloudConfigController struct {
+	infraClient     configv1client.InfrastructureInterface
+	infraLister     configv1listers.InfrastructureLister
+	configMapClient corev1client.ConfigMapsGetter
+
+	// transformers stores per platform tranformer
+	transformers map[configv1.PlatformType]transformer
+}
+
+// NewController returns a KubeCloudConfigController
+func NewController(operatorClient operatorv1helpers.OperatorClient,
+	infraClient configv1client.InfrastructuresGetter, infraLister configv1listers.InfrastructureLister, infraInformer cache.SharedIndexInformer,
+	configMapClient corev1client.ConfigMapsGetter,
+	recorder events.Recorder) factory.Controller {
+	c := &KubeCloudConfigController{
+		infraClient:     infraClient.Infrastructures(),
+		infraLister:     infraLister,
+		configMapClient: configMapClient,
+		transformers: map[configv1.PlatformType]transformer{
+			configv1.AWSPlatformType: awsTransformer,
+		},
+	}
+	return factory.New().
+		WithInformers(
+			operatorClient.Informer(),
+			infraInformer,
+		).
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		ResyncEvery(time.Minute).
+		ToController("KubeCloudConfigController", recorder)
+}
+
+func (c KubeCloudConfigController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	obj, err := c.infraLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		syncCtx.Recorder().Warningf("KubeCloudConfigController", "Required infrastructures.%s/cluster not found", configv1.GroupName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	currentInfra := obj.DeepCopy()
+	var platformName configv1.PlatformType
+	if pstatus := currentInfra.Status.PlatformStatus; pstatus != nil {
+		platformName = pstatus.Type
+	}
+	if platformName == "" {
+		syncCtx.Recorder().Warningf("KubeCloudConfigController", "Falling back to deprecated status.platform because infrastructures.%s/cluster status.platformStatus.type is empty", configv1.GroupName)
+		platformName = currentInfra.Status.Platform
+	}
+
+	sourceCloudConfigMap := currentInfra.Spec.CloudConfig.Name
+	sourceCloudConfigMapKey := currentInfra.Spec.CloudConfig.Key
+	if sourceCloudConfigMapKey == "" {
+		sourceCloudConfigMapKey = targetConfigKey // default to the known key
+	}
+	source := &corev1.ConfigMap{}
+	if len(sourceCloudConfigMap) > 0 {
+		obj, err := c.configMapClient.ConfigMaps(operatorclient.GlobalUserSpecifiedConfigNamespace).Get(ctx, sourceCloudConfigMap, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		obj.DeepCopyInto(source)
+		source.ObjectMeta = metav1.ObjectMeta{}
+	}
+
+	t, ok := c.transformers[platformName]
+	if !ok {
+		t = asIsTransformer
+	}
+
+	target, err := t(source, sourceCloudConfigMapKey, currentInfra)
+	if err != nil {
+		return err
+	}
+
+	targetCloudConfigMap := targetConfigName
+	targetCloudConfigMapKey := sourceCloudConfigMapKey
```

Comment:
I prefer being prescriptive here.  I don't see a reason why we can't be.  If there is one, please add a comment.

Reply:
hmm.. i'm not completely sure what you mean by the comment.

the key in the configmap that contains the cloud.conf is controlled by the sourceconfigmapkey. if there is no source, but we need a target config map we will use the _controlled controlled default key_ see https://github.com/openshift/cluster-config-operator/pull/120/files/650826a9518b5890af3b43ed27e1262ba53e2553..fe8c44193ee5245e7c735a94f81bf16099fa3559#diff-99d1c645e7848874c8b0c0ec171e00e1R85-R88

do you mean we should enforce that the key should always be `cloud.conf` ? that would have to be part of the infra api right??